arrow-ank/src/main/kotlin/arrow/ank/ank.kt

```diff
@@ -30,59 +24,45 @@ fun Long.humanBytes(): String {
   return String.format("%.1f %sB", this / unit.toDouble().pow(exp.toDouble()), pre)
 }
 
-fun <A> toValidatedNel(a: Either<Throwable, A>): ValidatedNel<Throwable, A> =
-  a.fold({ e -> Validated.Invalid(NonEmptyList(e)) }, { a -> Validated.Valid(a) })
+suspend fun ank(source: Path, target: Path, compilerArgs: List<String>, ankOps: AnkOps): Unit = with(ankOps) {
+  printConsole(colored(ANSI_PURPLE, AnkHeader))
+  val heapSize = Runtime.getRuntime().totalMemory()
+  val heapMaxSize = Runtime.getRuntime().maxMemory()
+  printConsole("Current heap used: ${(heapSize - Runtime.getRuntime().freeMemory()).humanBytes()}")
+  printConsole("Starting ank with Heap Size: ${heapSize.humanBytes()}, Max Heap Size: ${heapMaxSize.humanBytes()}")
+  val path = createTargetDirectory(source, target)
 
-fun <F> Concurrent<F>.ank(source: Path, target: Path, compilerArgs: List<String>, ankOps: AnkOps): Kind<F, Unit> = with(ankOps) {
-  fx.concurrent {
-    !effect { printConsole(colored(ANSI_PURPLE, AnkHeader)) }
-    val heapSize = Runtime.getRuntime().totalMemory()
-    val heapMaxSize = Runtime.getRuntime().maxMemory()
-    !effect { printConsole("Current heap used: ${(heapSize - Runtime.getRuntime().freeMemory()).humanBytes()}") }
-    !effect { printConsole("Starting ank with Heap Size: ${heapSize.humanBytes()}, Max Heap Size: ${heapMaxSize.humanBytes()}") }
-    val path = !effect { createTargetDirectory(source, target) }
-    val validatedPaths = !effect {
-      path.ankFiles().map { (index, p) ->
-        effect {
-          val totalHeap = Runtime.getRuntime().totalMemory()
-          val usedHeap = totalHeap - Runtime.getRuntime().freeMemory()
-          val message = "Ank Compile: [$index] ${path.relativize(p)} | Used Heap: ${usedHeap.humanBytes()}"
-          printConsole(colored(ANSI_GREEN, message))
-          val preProcessed = p.processMacros()
-          val (processed, snippets) = extractCode(preProcessed)
-          val compiledResult = compileCode(p toT snippets, compilerArgs)
-          val result = replaceAnkToLang(processed, compiledResult)
-          generateFile(p, result)
-        }.attempt().map(::toValidatedNel)
-      }.toList()
+  val validatedPaths = path.ankFiles().fold(listOf<ValidatedNel<Throwable, Path>>()) { acc, file ->
+    val res = try {
```

Comment:
tbh the only think I don't really like about all this is the usage of try-catch, apart from that the rest looks much better!

Reply:
@nomisRev and I discussed a bit this morning on the chance to move this to be Streams which are able to raise errors, but since I need this change for my workshop I opted for a direct port of what we had but using suspend + unwrapped style. I also think it makes more sense than streams, given this is not like a multiple emission problem.

Do you have any other idea to improve this with Arrow Fx Coroutines?

Reply:
We can refactor this to use emit `Validated` inside `Sequence` instead of throwing, but I'm not sure if we're going to continue using Ank. The new Dokka architecture allows for plugins, and KotlinX also has similar project to Ank already now that will probably integrate with the new Dokka.

Reply:
Can you review again after 27d2743 @nomisRev @aballano?